Fix for games with altered file extensions

sorlok opened the pull request:
Ready to merge (from my point of view). 

This is my implementation for Issue #1920; please see the initial discussion there.

Regarding @Ghabry's comments:
1) Check only 2 files: Done, but see note
2) Assume larger==LDB: Done
3) LcfXXX checks: Done, but see note

Some notes on this fix:
1) What's the easiest way to cheaply check if a directory *might* be a game project? Right now I check for RPG_RT.ini, but as you said that's not always present. I expect RPG_RT.exe might also be stripped. Is there any good file to check, or should I just disable this minor performance hack?  ---Side note: it might be worth keeping if *all games with hacked extensions* have .ini files.
2) The Meta file now has "LmtHeaderString=XXX", which allows us to pass the "XXX" string in to liblcf (i.e., to tell it what string to expect within the file header). As you said, errors are disabled if this string doesn't match  --but it still shows a warning. Do we want to allow passing in the expected header string to liblcf when parsing to silence the warning, or is this not worth the trouble?
3) We cannot apply the Meta file until *after* scanning the directory. Main reason: the Meta instance is stored in player.cpp. Sub reason: we don't know which file to calculate the CRC32 of until we scan (since they were renamed). This is not a huge problem, since the scanning is pretty fast.
4) I scan the names of all mapXXXX.ext files to ensure the extension is consistent across files. This might be overkill; we could (a) just stop on the first file or (b) check for the starting map filename (pulled from LMT), assuming we've parsed LMT at this point (which I think we have).
5) This fix assumes all "rpg_rt.XXX" files are renamed by extension ONLY --i.e., if you rename it to "abc.lmt" then this fix won't find it. Do any hacks do this, and do we care?
6) Just reiterating that the meta file can be used to overcome most limitations as discussed (e.g., slow lmu scanning and small LDB files).

Ghabry commented:
One quick comment: std::regex is broken on some platforms and can't be used :(

fdelapena commented:
Jenkins bot: ok to test

fdelapena commented:
Jenkins: test this please

fdelapena commented:
jenkins: test this please

Ghabry commented:
Thanks for your contribution. Really cool to have support for this. Reviewing can currently take longer because it is summer and vacation time, so don't worry if there is less feedback . We are more active during the cold months ;).

Ghabry commented:
(I was mentioning std::regex before but this should work properly since gcc5.0 so can be ignored)

- What does "Snh" mean?

> What's the easiest way to cheaply check if a directory might be a game project? Right now I check for RPG_RT.ini, but as you said that's not always present. I expect RPG_RT.exe might also be stripped.

When a game tree is parsed the file list is already in memory so the expensive operation (Read directory and alloc all the strings) is already done. I don't think doing some string testing will be really noticable. Also your code will only execute for the "not a game" case which shouldn't happen too often so don't optimize too much.

Sometimes games rename both RPG_RT and the ini. When the Exe is renamed it also changes the ini file it searches for (Game.exe will search for Game.ini)

> [LCF Header] As you said, errors are disabled if this string doesn't match --but it still shows a warning. Do we want to allow passing in the expected header string to liblcf when parsing to silence the warning, or is this not worth the trouble?

The "warning" is only printed on terminal so I don't consider this important

> I scan the names of all mapXXXX.ext files to ensure the extension is consistent across files.

I agree on you this is overkill. Just using the first mapfile you can find is good enough, we can still adopt if any game ever breaks this heuristic... (because there is some Map0001.bak there)

> [RPG_RT prefixed changed] Do any hacks do this, and do we care?

Not sure if any game does this. Is also a bit hard to analyse. When you provide me with some script I can use to scan hundreds of games easily I can tell you quickly ;)

sorlok commented:
> What does "Snh" mean?
Just my initials; I usually prefix branches with this to minimize the chance of conflict. Let me know if you prefer an alternative naming scheme (I can file a new PR).

> I don't think doing some string testing will be really noticeable.
Great, I removed the INI check.

> The "warning" is only printed on terminal so I don't consider this important
Great, I removed this from the sample .ini file

> Just using the first mapfile you can find is good enough
Done

> When you provide me with some script I can use to scan hundreds of games easily I can tell you quickly
That's not a bad idea. I'll write a quick Python script that can check this stuff.

sorlok commented:
Rebased and squashed. This branch is ready for merge (from my point of view --let me know if anything needs to change).

sorlok commented:
Hey all,

Just got back from vacation; wanted to check if there's anything more needed with this branch. (Also if you are all still on vacation that's fine too.) Thanks!

Ghabry commented:
I'm back from vacation and currently prepare running your script (extracting all the games takes a while \*g\*)

Ghabry commented:
Some games are molebox encrypted or have corrupted zip paths (\ instead of /) but for ones that work (around 1900): Only 2 are affected by this PR:

- Beyond Black Horizon (https://rmarchiv.de/games/443) - This has changed file extensions
- Verloren (https://rmarchiv.de/games/1333) - Uses default extensions but the maps are in a folder "MM" with names 0001.lmu etc. _out of scope_

sorlok commented:
> I'm back from vacation and currently prepare running your script (extracting all the games takes a while *g*)
> 
> Minor patch (recurses in a subdirectory when the subdir only has one folder)
> 
> ```
> 91,96c91
> <   dirlist = list(os.listdir(full_dir))
> <   dirs = list(filter(lambda x: os.path.isdir(os.path.join(full_dir, x)), dirlist))
> <   if len(list(dirs)) == 1:
> <     return scan_project(full_dir, dirs[0])
> < 
> <   for filepath in dirlist:
> ---
> >   for filepath in os.listdir(full_dir):
> ```

Updated, thanks

Ghabry commented:
This needs now a minor rebase

Ghabry commented on src/filefinder.cpp:
```diff
@@ -739,6 +741,132 @@ bool FileFinder::IsEasyRpgProject(DirectoryTree const& dir){
+		if(std::regex_search(item.first, m, rgx)) {
+			std::string ext = m[1].str();
+			if (ext != "exe" && ext != "ini") {
+				candidates.push_back({item.second, ext, GetFileSize(item.second)});
```
this doesn't work when launched from the Gamebrowser (wrong relative path).

Use:
```
candidates.push_back({item.second, ext, GetFileSize(FileFinder::FindDefault(dir, item.second))});
```

sorlok replied:
Fixed in 9e361a9bf1052, thanks

Ghabry commented:
In my first quick test it worked. Good job :)

sorlok commented:
Rebased; I'm running a few quick tests (and waiting on builds).

sorlok commented:
Tested again on Linux and Android; confirmed it's working.

Ghabry commented on src/filefinder.cpp:
```diff
@@ -749,6 +751,132 @@ bool FileFinder::IsEasyRpgProject(DirectoryTree const& dir){
+	return !GetRPG2kProjectWithRenames(dir).Empty();
+}
+
+FileFinder::RPG2KNonStandardFilenameGuesser FileFinder::GetRPG2kProjectWithRenames(DirectoryTree const& dir) {
```
Can you move the entire guess/remap logic in a new file? My problem here is that I have a huge PR touching FileFinder in the works and such a huge change is terrible to handle for me :(.

No idea. Make a name up: "filefinder_guesser.cpp/h", "file_guesser.cpp/h", "project_guesser.cpp/.h".

Move ``GetRPG2kProjectWithRenames, GuessAndAddLmuExtension`` and ``RPG2KNonStandardFilenameGuesser, RPG2KFileExtRemap`` to this

``IsRPG2kProjectWithRenames`` can stay as it fits with the rest.

sorlok replied:
Done.

sorlok commented on builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameScanner.java:
```diff
@@ -128,12 +132,26 @@ private static boolean isRpg2kGame(File dir) {
+        if (rpgRTcount == 2) {
+            return true;
+        }
+
```
Done.

sorlok commented on builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameBrowserHelper.java:
```diff
@@ -50,6 +50,9 @@ public static boolean isRpg2kGame(File dir) {
         boolean treemapFound = false;
 
+        // Create a lookup by extension as we go, in case we are dealing with non-standard extensions.
+        int rpgRTcount = 0;
```
Done.

sorlok commented on src/game_map.cpp:
```diff
@@ -271,17 +271,14 @@ void Game_Map::SetupCommon(int _id, bool is_load_savegame) {
-	ss << "Map" << std::setfill('0') << std::setw(4) << location.map_id << ".emu";
-
-	std::string map_file = FileFinder::FindDefault(ss.str());
+	std::string mapName = Game_Map::ConstructMapName(location.map_id, true);
```
Done.

sorlok commented on src/game_map.cpp:
```diff
@@ -271,17 +271,14 @@ void Game_Map::SetupCommon(int _id, bool is_load_savegame) {
 
 		if (map_file.empty()) {
-			Output::Error("Loading of Map {} failed.\nThe map was not found.", ss.str());
+			Output::Error("Loading of Map {} failed.\nThe map was not found.", mapName.c_str());
```
Done, thanks for the info (I don't know much about this output api).

sorlok commented on src/game_map.cpp:
```diff
@@ -1703,11 +1701,18 @@ int Game_Map::GetTargetPanY() {
 }
 
-FileRequestAsync* Game_Map::RequestMap(int map_id) {
+std::string Game_Map::ConstructMapName(int map_id, bool isEasyRpg) {
```
Done.

sorlok commented on src/options.h:
```diff
@@ -48,13 +48,26 @@
-#define DATABASE_NAME "RPG_RT.ldb"
-#define DATABASE_NAME_EASYRPG "EASY_RT.edb"
+#define DATABASE_NAME RPG_RT_PREFIX "." SUFFIX_LDB
+#define DATABASE_NAME_EASYRPG RPG_RT_PREFIX "." SUFFIX_EDB
```
Fixed. Sorry about that, it was not intentional.

sorlok commented:
Changes made and rebased. Tested and works on Linux. I'm still building on Android.

sorlok commented:
Android is working great. That's everything on my side.

Ghabry commented on src/filext_guesser.h:
```diff
@@ -0,0 +1,97 @@
+
+}
+
+#endif
```
new line at end of file missing

sorlok replied:
Fixed.

Ghabry commented on src/filext_guesser.h:
```diff
@@ -0,0 +1,97 @@
+/**
+ * FilextGuesser contains helper methods for guessing the extensions used on non-standard RPG Projects.
+ */
+namespace FilextGuesser {
```
I think there is a typo here. Should be FileExtGuesser and the file fileext_guesser.

sorlok replied:
Hmm, I was thinking "filext", but I now think that might be trademarked by the associated ".com" address. I'll change this after work.

Ghabry reviewed (commented):
Except for my remaining nits this looks now fine for me.

Do you know how to "squash" in a rebase? If yes please squash the two "Code review" commits into one.

I marked it as "Has PR deps" because I want to merge this after #2293
(that PR changes some string functions and want to attempt a regex -> starts/ends_with change as I expect this to be faster here)

sorlok commented:
Renamed filext to fileext (and similar).
Switched from regex to string_view.
Rebased, squashed.

Code seems like it's still working (tested loading two games on Linux).

Ghabry commented on src/filefinder.cpp:
```diff
@@ -26,6 +26,7 @@
 #include <string>
 #include <vector>
 #include <sstream>
+#include <regex>
```
unnecessary include (thanks for changing this :+1: )

sorlok replied:
Fixed (and you're welcome!)

sorlok replied:
Also fixed the commit message.

Ghabry reviewed (commented):
Want to quickly test this but it looks now fine for me! Thanks! Very useful feature :)

The commit message can be improved. Instead of "Squashed commits" can you use something like "Support games with altered file extensions" (you can keep the verbose message part as is)

Ghabry reviewed (changes requested):
I planned to merge it but I just found an important bug (sorry).

When starting a renamed game the output is "Debug: Encoding not detected". This means that for non-English games the language detection is broken (and they won't run). The encoding detection still only uses RPG_RT.ldb.

I think you must reorder the code a bit. The following must happen before "GetEncoding" is called:

1. Load Meta file
2. Guess extensions
3. GetEncoding
4. [...] stuff
5. LoadDatabase

sorlok commented:
Ah, interesting. Glad you caught this bug before merging! I'm a bit busy this weekend, but I'll check it out next week.

Ghabry commented:
About the build failures: That was a very bad timing for updating the PR. There is currently a breaking change from liblcf pending/rebuilding.
You will have to rebase again in ~1h ^^'.

sorlok commented:
> About the build failures: That was a very bad timing for updating the PR. There is currently a breaking change from liblcf pending/rebuilding.
> You will have to rebase again in ~1h ^^'.

No worries; I'm still working on the fix.

sorlok commented:
Ok, the fix is in (tested on Laxius Power and a standard game).
1) I left it as its own commit for now, since I think that makes it easier to review. Let me know if/when you want it squashed.
2) Are there any EasyRPG format games I can test this on? I want to make sure I didn't break anything with the new format.

Ghabry reviewed (approved):
This works perfect now. Tested it. Thanks for this great contribution!

Ghabry commented:
@sorlok 
FYI
Made an announcement: https://twitter.com/EasyRPG/status/1301108541185548288

sorlok commented:
Cool, thanks!
